### PR TITLE
add bloc

### PR DESCRIPTION
**Pull request by alexei-kruk-idf**

### CODE REVIEW

**Review by yuliya-klescheva-idf — APPROVED** [review]

File: packages/domain/lib/use_cases/interfaces/palindrome_case.dart [hunk 1]
```diff
@@ -1,3 +1,3 @@
 abstract class PalindromeCase {
-  bool isPalindrome(String text);
+  Future<bool> call(String text);
```
Comment (sergei-kabak-idf) [inline packages/domain/lib/use_cases/interfaces/palindrome_case.dart]: Why do you use Future <bool> call () ?

**Review by sergei-kabak-idf — CHANGES_REQUESTED** [review]

File: packages/domain/lib/use_cases/interfaces/palindrome_case.dart (hunk 1 above)
Comment (alexei-kruk-idf) [reply packages/domain/lib/use_cases/interfaces/palindrome_case.dart]: fixed

**Review by sergei-kabak-idf — APPROVED** [review]

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import 'dart:async';
+
+import 'package:domain/use_cases/impl/palindrome_impl.dart';
+
+class PalindromeBlock {
+  final PalindromeCaseImpl _palindromeUseCase;
+
+  PalindromeBlock(this._palindromeUseCase) {
+    _inputStreamController.stream.listen(_checkPalindrome);
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: why do u need the stream here to pass events from widget to bloc if u can call bloc methods directly?

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import 'dart:async';
+
+import 'package:domain/use_cases/impl/palindrome_impl.dart';
+
+class PalindromeBlock {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: Block -> Bloc, but it's ok :)

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  Stream<String> get outputStreamSink => _outputStreamController.stream;
+
+  void dispose() {
+    _inputStreamController.close();
+    _outputStreamController.close();
+  }
+
+  void _checkPalindrome(String text) {
+    final isPalindrome = _palindromeUseCase(text);
+    var textResult = text.trim();
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: ??? why do u need it

File: packages/presentation/lib/home_screen/screen/home_widget.dart [hunk 5]
```diff
@@ -9,39 +10,49 @@ class HomeWidget extends StatefulWidget {
 }
 
 class _HomeWidgetState extends State<HomeWidget> {
-  final palindromeCaseImpl = PalindromeCaseImpl();
+  final palindromeBlock = PalindromeBlock(PalindromeCaseImpl());
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/home_screen/screen/home_widget.dart]: bloc

File: packages/presentation/lib/home_screen/screen/home_widget.dart [hunk 6]
```diff
@@ -9,39 +10,49 @@ class HomeWidget extends StatefulWidget {
...
-        Text(textResult)
+        StreamBuilder(
+            stream: palindromeBlock.outputStreamSink,
+            builder: (context, snapshot) {
+              if (snapshot.data != null) {
+                return Text(snapshot.data as String);
+              } else {
+                return Text('');
+              }
+            })
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/home_screen/screen/home_widget.dart]: in most cases there's only one stream which is used to communicate via bloc and widget. in this case u should have the following widget hierarchy:

StreamBuilder
|
|- Scaffold

no need to change it now, but use this approach while doing a bigger project

**Review by sergei-mikhailovskii-idf — APPROVED** [review]

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart (hunk 3 above)
Comment (alexei-kruk-idf) [reply packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: fixed

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart (hunk 4 above)
Comment (alexei-kruk-idf) [reply packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: fixed

File: packages/presentation/lib/home_screen/screen/home_widget.dart (hunk 5 above)
Comment (alexei-kruk-idf) [reply packages/presentation/lib/home_screen/screen/home_widget.dart]: fixed

File: packages/presentation/lib/home_screen/screen/home_widget.dart (hunk 6 above)
Comment (alexei-kruk-idf) [reply packages/presentation/lib/home_screen/screen/home_widget.dart]: okay, i'll be know

File: packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart (hunk 2 above)
Comment (alexei-kruk-idf) [reply packages/presentation/lib/home_screen/bloc/palindrome_bloc.dart]: fixed

**Pull request merged**